src/back/msl/writer.rs

```diff
@@ -786,20 +786,57 @@ impl<W: Write> Writer<W> {
 
     fn put_atomic_fetch(
         &mut self,
-        pointer: Handle<crate::Expression>,
+        expr_handle: Handle<crate::Expression>,
         key: &str,
         value: Handle<crate::Expression>,
         context: &ExpressionContext,
     ) -> BackendResult {
-        write!(
-            self.out,
-            "{}::atomic_fetch_{}_explicit({}",
-            NAMESPACE, key, ATOMIC_REFERENCE
-        )?;
-        self.put_expression(pointer, context, true)?;
-        write!(self.out, ", ")?;
-        self.put_expression(value, context, true)?;
-        write!(self.out, ", {}::memory_order_relaxed)", NAMESPACE)?;
+        // TODO Why not always put parentheses?
+        let is_scoped = true;
+        // Atomic ops are structured different from a normal load.
+        // Rewrite this for the bounds-check case so that the bounds check happens once
+        // when resolving the pointer, and then use an unchecked access within the atomic expression itself.
+        let policy = context.choose_bounds_check_policy(expr_handle);
+        if policy == index::BoundsCheckPolicy::ReadZeroSkipWrite
+            && self.put_bounds_checks(
+                expr_handle,
+                context,
+                back::Level(0),
+                if is_scoped { "" } else { "(" },
+            )?
+        {
+            write!(self.out, " ? ")?;
+
+            {
+                // True branch of bounds check ternary.
+                write!(
+                    self.out,
+                    "{}::atomic_fetch_{}_explicit({}",
+                    NAMESPACE, key, ATOMIC_REFERENCE
+                )?;
+                self.put_access_chain(expr_handle, policy, context)?;
+                write!(self.out, ", ")?;
+                self.put_expression(value, context, true)?;
+                write!(self.out, ", {}::memory_order_relaxed)", NAMESPACE)?;
+            }
+
+            write!(self.out, " : DefaultConstructible()")?;
+
+            if !is_scoped {
+                write!(self.out, ")")?;
+            }
+        } else {
+            write!(
+                self.out,
+                "{}::atomic_fetch_{}_explicit({}",
+                NAMESPACE, key, ATOMIC_REFERENCE
+            )?;
+            self.put_expression(expr_handle, context, true)?;
```

Comment:
This is a pre-existing problem, but, this should call `put_access_chain` instead of `put_expression`, for consistency with the atomic case in `put_unchecked_load`. That also won't check again for a `ReadZeroSkipWrite` policy, which at this point we already know does not apply. I would expect this to have no effect on the generated code.

Reply:
Factored it into a closure.